tests/labels_test.go

```diff
@@ -0,0 +1,122 @@
+	resource := res.NewResource()
+	testExpectedLabels(resource, res.GetKey(), res.ExpectedLabels)
+}
+
+func expectAppLabelsRestoreAfterUpdate(res *testResource) {
+	testExpectedLabelsRestoreAfterUpdate(res.NewResource(), res.GetKey(), res.ExpectedLabels)
+}
+
+func testExpectedLabels(resource controllerutil.Object, key client.ObjectKey, expectedLabels map[string]string) {
+	patchAppLabelsIntoSSP(expectedLabels)
+	waitForLabelMatch(resource, key, expectedLabels)
+}
+
+func testExpectedLabelsRestoreAfterUpdate(resource controllerutil.Object, key client.ObjectKey, expectedLabels map[string]string) {
+	testExpectedLabels(resource, key, expectedLabels)
+
+	operations := newLabelOperations(expectedLabels)
+	for label := range expectedLabels {
+		operations = append(operations, labelPatchOperationFor("replace", label, "wrong"))
+	}
+	patch := encodePatch(operations)
+	err := apiClient.Patch(context.TODO(), resource, patch)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(labelsMatch(expectedLabels, resource)).To(BeFalse())
+}
+
+func patchAppLabelsIntoSSP(expectedLabels map[string]string) {
+	ssp := getSsp()
+	Expect(ssp).NotTo(BeNil())
+
+	if ssp.Labels != nil && labelsMatch(expectedLabels, ssp) {
+		return
+	}
+
+	newLabels := map[string]string{
+		common.AppKubernetesPartOfLabel:  expectedLabels[common.AppKubernetesPartOfLabel],
+		common.AppKubernetesVersionLabel: expectedLabels[common.AppKubernetesVersionLabel],
+	}
+
+	patch := buildLabelsPatchAdding(newLabels)
+	err := apiClient.Patch(context.TODO(), ssp, patch)
+	Expect(err).NotTo(HaveOccurred(), "app labels could not be added to SSP CR")
+}
+
+func buildLabelsPatchAdding(labels map[string]string) client.Patch {
+	return buildLabelsPatch("add", labels)
+}
+
+func buildLabelsPatch(operation string, labels map[string]string) client.Patch {
+	operations := newLabelOperations(labels)
+
+	for label, value := range labels {
+		operations = append(operations, labelPatchOperationFor(operation, label, value))
+	}
+
+	return encodePatch(operations)
+}
+
+func newLabelOperations(labels map[string]string) []jsonpatch.Operation {
+	operations := make([]jsonpatch.Operation, 0, len(labels)+1)
+	operations = append(operations, jsonpatch.NewPatch("add", "/metadata/labels", struct{}{}))
+	return operations
+}
+
+func labelPatchOperationFor(op, label, value string) jsonpatch.Operation {
+	return jsonpatch.NewPatch(op, fmt.Sprintf("/metadata/labels/%s", strings.ReplaceAll(label, "/", "~1")), value)
+}
+
+func encodePatch(operations []jsonpatch.Operation) client.Patch {
+	patchBytes, err := json.Marshal(operations)
+	Expect(err).NotTo(HaveOccurred())
+
+	fmt.Fprintf(GinkgoWriter, "sending patch: %s", string(patchBytes))
+	return client.RawPatch(types.JSONPatchType, patchBytes)
+}
+
+func waitForLabelMatch(resource controllerutil.Object, key client.ObjectKey, expectedLabels map[string]string) {
+	Eventually(func() bool {
+		err := apiClient.Get(context.TODO(), key, resource)
+		if err != nil {
```

Comment:
I think the test can expect that the `err` is `nil`, we do it in other places. The resource exists, and if the `Get()` call returns an error, it is probably some unexpected issue.

Reply:
I'm not sure I understand what you mean.
Right now it expects `nil` but if it gets and error it retries until it's fine again.

Reply:
Oh, you mean Assert it is nil.
Yeah this seems more resilient, but I can switch it if you like.